### DISCUSSION TITLE
The future of Widgets

### DISCUSSION THREAD

**Discussion by tomek-szczesny** [dankamongmen/notcurses#1893] [Ideas]
So me and @chromafunk been talking, and although he did not commit to any of that just yet, I'm curious whether we have any "rules" about which widgets are welcomed in notcurses and which are not.

The braille-based graph plotter that @chromafunk has found implemented in Python, could be a perfect widget for notcurses - even better, utilizing nc's internal possibilities it could be much faster and not limited to Braille symbols.

Personally I am interested in some clever ways of representing scientific data for my other project. Widgets with graphs, heat maps, multi-layer vertical progress bar arrays, pie charts or whatever.

I trust you've got nothing against features like that. Perhaps it's a high time to organize widgets somehow. Right now they sit mixed up with "core" features of notcurses, and when notcurses gets its well deserved recognition, I suspect the widget collection will explode.
I suspect there will be a demand for organizing widgets both within the code, and from the user's perspective. A basic set of rules how to contribute a widget, and perhaps a naming convention surely won't hurt. 

Or maybe the higher level widgets are not to be a part of notcurses directly, that is something that should be decided at some point as well.

**Comment by journeytosilius** [comment]
I'm up for this, the sooner the better. Since I either need to: 
A- make a module for my software with what we got
B- do it the right way and create a widget that can be used straight from notcurses, it's fast and reusable

**Reply by tomek-szczesny** [reply]
Well, I think you can start coding it anytime. In the best case scenario @dankamongmen will fully support your work and help you out, and in the worst case you'll create a separate library dependent on notcurses :p

I think you should start by studying the existing widget, can't remember its name... The one that shows fps graph on the bottom of notcurses-demo.

**Reply by journeytosilius** [reply]
cool, I'll start taking a look in to it asap then, keep you updated

**Comment by dankamongmen** [comment]
> So me and @chromafunk been talking, and although he did not commit to any of that just yet, I'm curious whether we have any "rules" about which widgets are welcomed in notcurses and which are not.
> 
> The braille-based graph plotter that @chromafunk has found implemented in Python, could be a perfect widget for notcurses - even better, utilizing nc's internal possibilities it could be much faster and not limited to Braille symbols.

i'm all for more widgets, and they're one area that i really hope will bring more developers to Notcurses. @MasFlam put together a lovely tabbed interface widget.

i'd prefer to keep them in-library if possible. even if they have some kind of new dependency, i'm happy to write the necessary CMake code to allow that dep to be optional, leaving out said widget if it's not there. of course, if someone wants to release them as their own library, that's absolutely fine (and might be better, since that way you can depend on the widget library and know you'll have it, as opposed to deping on libnotcurses-core by itself and not being certain).

my only requirement would be that they ought work like the other widgets, unless it's absolutely impossible:
* they ought usually accept ownership of an `ncplane` on which to do their thing rather than creating their own
  * this plane ought be destroyed on all error paths in the constructor, and in the destructor
* any subplanes they need create ought be properly bound so that the widget moves as a single entity
* each widget requires a man page and unit tests
  * and ideally also `USAGE.md` updates and even example code
* reuse as much code as possible

**Reply by dankamongmen** [reply]
> i'd prefer to keep them in-library if possible. even if they have some kind of new dependency, i'm happy to write the necessary CMake code to allow that dep to be optional, leaving out said widget if it's not there. of course, if someone wants to release them as their own library, that's absolutely fine (and might be better, since that way you can depend on the widget library and know you'll have it, as opposed to deping on libnotcurses-core by itself and not being certain).

now, i might have been a bit too permissive here -- i'm not bringing in python deps that we then wrap in c, etc. that's a bit much. however, for this example, expanding the `ncplot` concept is something i've had plans to do for some time. i'd love to see linegraphs at a minimum. but we'd almost certainly be writing our own code (likely using lots of stuff factored out of `plot.c`), and definitely not calling into some Python module.

**Reply by dankamongmen** [reply]
and i'm happy to prioritize stuff people communicate a real-world need for. right now i have the following goals for 2.3.9: https://github.com/dankamongmen/notcurses/discussions/1851. we release roughly once a week. so linegraphs could easily come on-schedule this month.

the one thing i'm generally *not* going to do is merge something inherently sloppy just to get feature coverage. i don't mind moving in pieces, but each one needs to be solid.

i can also say that as more developers come on board, and they write code i'm not using regularly and directly, it's going to be absolutely critical that they write unit tests (and ideally a little integration example as a PoC).

right now i'm mainly focused on wrapping up the kitty animation protocol work, which might happen tonight (fingers crossed), and then my big thing is the new scrolling-cli-in-rendered-mode concept that `notcurses-info` is proving out. those are absolutely critical core pieces, and (unlike periphery widgets) need to be done before abi3 can be declared. that second part might take a few days, or a few weeks; i'm unsure.

**Reply by tomek-szczesny** [reply]
> now, i might have been a bit too permissive here -- i'm not bringing in python deps that we then wrap in c, etc. that's a bit much. however, for this example, expanding the `ncplot` concept is something i've had plans to do for some time. i'd love to see linegraphs at a minimum. but we'd almost certainly be writing our own code (likely using lots of stuff factored out of `plot.c`), and definitely not calling into some Python module.

I agree, and we all understand notcurses are developed in C, thus any new widget shall be coded in C as well.
I had to learn C to start working on the VT thing, I'm getting used to it already. ;)

I take it as a warning you may be busy with more important stuff in the following weeks, and I don't think it's a major setback.
My model is to write the VT support in a completely separate program, and when it works well, book some of your time on a goals list and move the parts of the prototype into the library, one after another. We shall see how it works in the end, but at least I'm not nagging you with PRs that may or may not be necessary in the final thing, and I'm not working on a notcurses clone that will be thousand commits behind when I'm done.

**Reply by dankamongmen** [reply]
> I take it as a warning you may be busy with more important stuff in the following weeks, and I don't think it's a major setback.

hehehe, look over the history of the project; i've been busy since 2019-11 =]. 6k commits in about 600 days, you do the math.

like i said somewhere a few days ago, i'll usually either reply to something within an hour, or it will be a few days. i need to go get on the macmini that @michaelsbradleyjr set up like a boss and make good on that investment. but if you put something up, i'm not going to let it rot.

**Reply by tomek-szczesny** [reply]
> like i said somewhere a few days ago, i'll usually either reply to something within an hour, or it will be a few days. i need to go get on the macmini that @michaelsbradleyjr set up like a boss and make good on that investment. but if you put something up, i'm not going to let it rot.

Sure enough you're a busy guy. And it pays!

I don't know much about cooperative code development, but I can at least predict that with more contributors, your work may gradually shift into overseeing the work of the others, at least partially. ;)

My shitty PR is a fine example - thanks for your patience and useful remarks. I promise to contribute at least tested stuff in the future. ;)
The VT thing is getting near full usability. Colors work, UTF-8 works, next stop are cursor movements (when you merge the PR :P).

**Reply by journeytosilius** [reply]
yep, and one of the goals on this is to get rid of python. once you enter micro-service territory you can't afford 190mb docker images. it's like ... adding a 190mb container to fire a script that paints braille plots ... takes all the logic out of it :)

**Reply by journeytosilius** [reply]
by the way, I'm normally using this on my k8s clusters : https://github.com/derailed/k9s

**Comment by tomek-szczesny** [comment]
Speaking of Widgets,
Concerning the VT thing, over time I'm growing more convinced the VT should indeed be a widget, rather than just a putvt() function and a context struct.

Here are a few examples why I think this is the case, and I'd be happy to hear what you think.
- Terminals generally send data to applications, and may receive commands altering their default behavior in this regard ([Mouse support](https://github.com/tomek-szczesny/notcurses-vt-proto/issues/8), [arrow keys coding](https://github.com/tomek-szczesny/notcurses-vt-proto/issues/10#issuecomment-876706845))
- Terminals have scrollback buffers and alternate buffers, and quite complex states, such as scrolling regions. Sending anything else than putvt() to ncplane makes no sense and will surely break the consistency of the session,
- VT thus must provide callback pointers for terminal output, or maybe even stuff like BEL
- VT must have extra function to let a user [scroll through scrollback ](https://github.com/tomek-szczesny/notcurses-vt-proto/issues/5)
- Why do I keep using vim shortcuts on a github page

Surely enough, most of that could be handled with a few functions taking a VT context struct as an argument, but on the other hand, isn't that what makes widgets stand out from a regular ncplane?
And what's the point of tainting an ncplane API with a bunch of functions that are usable only in a very narrow, specific circumstances.

For now, I have no clue about widgets in general, and I can only imagine that widgets are the solution to the points I've risen here. I keep refactoring putvt() so it is clean-ish and maintainable, and supporting input-only operation.

**Reply by o-sdn-o** [reply]
In addition to the full duplex VT/text channel between the application and the terminal (STDIN/OUT), there is also a communication channel based on system callbacks. For example, tracking PTY size, and other signals.

**Reply by tomek-szczesny** [reply]
This has to stop somewhere.

From my perspective, the point of VT widget is to support visual and input portion of the terminal - what notcurses is all about. It may be a future improvement, maybe `ncsubprocess` or something, but for now I think it's up to a user to set up a stream from and to the program.

**Reply by o-sdn-o** [reply]
Many terminal apps refuse to render visual output if they cannot get the PTY size. Almost all __TUI/visual__ terminal apps.

**Reply by tomek-szczesny** [reply]
That's true, and this is something that, in my humble opinion, notcurses doesn't need to be bothered with. At least for now.
For example, I don't want notcurses to handle any pty business for me, because I got it covered by libssh. I just want an input and output buffer.

**Reply by o-sdn-o** [reply]
Once everything works through libssh, it will be very easy to switch to system PTYs in the future. 🙂

**Reply by tomek-szczesny** [reply]
I certainly hope so, because I might not be the right person to do that. :)

**Reply by dankamongmen** [reply]
> In addition to the full duplex VT/text channel between the application and the terminal (STDIN/OUT), there is also a communication channel based on system callbacks. For example, tracking PTY size, and other signals.

yep, both `ioctl(2)` calls and signals (particularly `SIGWINCH`)

**Reply by dankamongmen** [reply]
yeah i don't see the active management aspects as relevant until we get to `ncsubprocess` filtered through this (where they *will* be very relevant).

**Reply by dankamongmen** [reply]
yeah keep working on what you're doing, @tomek-szczesny . full integration with `ncsubprocess` will require deep UNIX knowledge and i'd rather handle it

**Reply by tomek-szczesny** [reply]
And concerning the original topic of this thread, I'll take it as a "sure, whatever"

**Reply by dankamongmen** [reply]
> And concerning the original topic of this thread, I'll take it as a "sure, whatever"

i am eagerly looking forward to the VT core. widgets, widgets i find it hard to get excited about. btw don't you have a PR out i need to go approve? let me do that.

**Reply by tomek-szczesny** [reply]
> i am eagerly looking forward to the VT core. widgets, widgets i find it hard to get excited about. btw don't you have a PR out i need to go approve? let me do that.

Thanks, I needed that today.

As I stated above, I don't see ncplane_putvt() as a viable option anymore. Terminal context turned out to be so complex that ncplane_putvt() would have to be used de facto exclusively, as any single stray input would ruin the thing.
That's why I turned into the idea of a widget instead, that accepts no other inputs than VT garbage.

I missed that request to keep parser and widget separate, but I do see potential use. In the end it seems that I'm literally rewriting libvterm right now.
So, you want a "VT core" that gets a stream of terminal nonsense, identifies escape sequences, and then what?
How do you envision a hypothetical VT widget interacting with VT core?
From my limited set of ideas, callbacks, I think?

**Reply by tomek-szczesny** [reply]
@dankamongmen , in another words, I need you to rephase what exactly you expect me to do. From what I gathered, the goal was to interpret VTspeak and put it on ncplane. The work in progress does that, including a state machine that applies direct actions on ncplane, once a complete sequence has been found. So, it's not exactly reusable anyway, be it a widget or not. 

If you prefer a separate VTspeak parser AND a separate widget that collects parser's output and applies actions to underlying ncplane, I could do that, but I need a tip or two how to approach this problem.
I think the solution has something to do with callbacks, but since there are no classes, planes or widgets don't have unique function instances that can handle callbacks unambiguously. Thus the callbacks would have to have a parameter that points at the struct in question. Maybe something like:

```
struct vtparser {
  char* buffer;
  int counter; // and all that internal stuff
  void* target;
  int (*sgr_4b_fg_cb)(void* payload, void* target);
  int (*sgr_4b_bg_cb)(void* payload, void* target);
  int (*sgr_8b_fg_cb)(void* payload, void* target);
  int (*sgr_8b_bg_cb)(void* payload, void* target);
  // And a metric tonne more
}
```

`int vt_parse(struct vtparser* p, const char* buf, size_t s);`

And on the VT widget side:

```
int vt_sgr_4b_fg_cb(void* value, void* vt_widget) {
  // cast 'value' into long*, or char[3]
  // cast vt_widget into vt*
  // do stuff
}
```

I'm not entirely sure if this is what you're after, the list of callbacks would be 100+ functions for sure. 
There could be one callback instead, with some enum type "sequence code", but then whatever function is being called, it will have to deal with all codes it may possibly get.

I would deeply appreciate your assistance, yo.

**Reply by dankamongmen** [reply]
hey sorry i missed this.

what you've written (as described in the first para) sound like exactly what i'd hope for:

> rom what I gathered, the goal was to interpret VTspeak and put it on ncplane. The work in progress does that, including a state machine that applies direct actions on ncplane, once a complete sequence has been found.

but that doesn't seem to imply

> So, it's not exactly reusable anyway, be it a widget or not.

like i'm imagining your interface is something like

```c
typedef struct ncvtctx {
 // whatever state you need across calls
} ncvtctx;

ncvtctx* ncvtctx_create(const char* term); // assume input was generated using TERM=term, or current if NULL
void ncvtctx_reset(ncvtctx* n); // reset lexing state, if you include this
void ncvtctx_destroy(ncvtctx* n);
int ncvtctx_feed(ncvtctx* n, ncplane* np, const char* b, size_t blen); // feed blen bytes of b to n, compiling it to np
```

so i call this with my output from wherever. i can pass the entirety of it in one call, or i can spoonfeed it a byte at a time. either way, as it chews input, it's performing output and state transitions on the `ncplane`.

**Reply by dankamongmen** [reply]
with this core, you can:

* open a file and pass it through all at once
* open a file and pass it through in chunks
* open a pipe and pass it through
* generate buffers locally and pass them through
* interleave a pipe/file with data you make up yourself
* send it random data, for fuzzing purposes
* launch a subprocess, read its output, and stream it

etc. it maps input bytes to an intermediate representation (realized however explicitly or unexplicitly you'd like) of DEC VT/ANSI concepts, and then to output functions on an ncplane, as they're completed. it carries just enough state to carry control sequences and multibyte UTF8 across calls. this is what i've thought you were hacking up for sure.

**Reply by tomek-szczesny** [reply]
> like i'm imagining your interface is something like

Yep, you are imagining it quite well and this is what is happening right now. I thought that ncvtctx_feed() that works only with ncplanes would not be seen as "reusable", as it cannot parse VTspeak for any other purpose than writing to ncplane.

**Reply by tomek-szczesny** [reply]
Also, it should be somehow guaranteed that ncplane has default settings and nobody dares to turn on autoscroll or write anything else to it. If it changes size, however, ncvtctx probably should know about that.

**Reply by tomek-szczesny** [reply]
> ```c
> ncvtctx* ncvtctx_create(const char* term); // assume input was generated using TERM=term, or current if NULL
> ```

I don't think that matters. ncvtctx will have its own terminfo entry and it's up to the data source to comply with its capabilities. One of the points of ncvtctx is to get stuff from the underlying program and translate to notcurses insternal format. I believe nc will do its job and spit out a result in correct format.

**Reply by dankamongmen** [reply]
hrmmmm i'm not sure about that last comment. like if someone moves the plane's virtual cursor between calls, i would expect absolute moves to be unchanged, but relative moves/non moves to be affected.

you may assume that all output is flowing through notcurses; that's a requirement of rendered mode. if that's violated, they need call `notcurses_refresh()` to get the frame redrawn *in toto*. but the user can definitely do other things to the plane between calls (the user may not call two plane-mutating functions on the same plane concurrently, see ["threads" from notcurses(3)](https://notcurses.com/notcurses.3.html).)

**Reply by dankamongmen** [reply]
> I don't think that matters. ncvtctx will have its own terminfo entry and it's up to the data source to comply with its capabilities. One of the points of ncvtctx is to get stuff from the underlying program and translate to notcurses insternal format. I believe nc will do its job and spit out a result in correct format.

hrmmm. there are good and bad things to say about this idea. even if this is the case, since you presuppose a notcurses terminfo entry, they could just call it with `("notcurses")` and this would all still work, while allowing us to be more flexible. i'm unsure why you would require the client to have data targeting a `notcurses` TERM. what if it was created before Notcurses existed?

**Reply by dankamongmen** [reply]
and i don't necessarily love the idea of a notcurses terminfo entry. a lot of capabilities are dynamic and dependent upon the environment. terminfo doesn't have any means of capturing and reflecting that.

**Reply by tomek-szczesny** [reply]
> hrmmmm i'm not sure about that last comment. like if someone moves the plane's virtual cursor between calls, i would expect absolute moves to be unchanged, but relative moves/non moves to be affected.
> 
> you may assume that all output is flowing through notcurses; that's a requirement of rendered mode. if that's violated, they need call `notcurses_refresh()` to get the frame redrawn _in toto_.

This is all true. I just don't think there is any meaningful way of putting anything to ncplane when it is already filled with ncplane_putvt(). But on the other hand, it's up to a user to use provided tools as intended.

I'm thinking about further consequences of stuff like ncplane scrolling feature - if set, I imagine some cursor moving and backbuffer functionalities would have to be handled differently. The question is, should I assume a default state of ncplane_scrolling, or prepare for both.

**Reply by dankamongmen** [reply]
i would imagine that this functionality would be useful in both a scrolling and non-scrolling situation. the former ought be easier for you, since you can continuously dump text and things just work, whereas with the latter case you must manually move to another line etc.

**Reply by tomek-szczesny** [reply]
> hrmmm. there are good and bad things to say about this idea. even if this is the case, since you presuppose a notcurses terminfo entry, they could just call it with `("notcurses")` and this would all still work, while allowing us to be more flexible.

I may understand your request wrong, but I don't think we're able to cover all terminals from terminfo. If we pass terminfo entry name to VT parser, what exactly should it do with that information?

May I remind you, `notcurses` terminfo entry was your idea, and I just agreed. ;) We may scrap that one, of course.

**Reply by tomek-szczesny** [reply]
> i would imagine that this functionality would be useful in both a scrolling and non-scrolling situation. the former ought be easier for you, since you can continuously dump text and things just work, whereas with the latter case you must manually move to another line etc.

Does that mean that I can let parser change that setting? Oh, but it can, and revert it to whatever was set before, once it does its thing. Shouldn't do any harm, I guess?

**Reply by tomek-szczesny** [reply]
@dankamongmen,

> I may understand your request wrong, but I don't think we're able to cover all terminals from terminfo. If we pass terminfo entry name to VT parser, what exactly should it do with that information?

**Reply by dankamongmen** [reply]
> @dankamongmen,
> 
> > I may understand your request wrong, but I don't think we're able to cover all terminals from terminfo. If we pass terminfo entry name to VT parser, what exactly should it do with that information?

it needn't do anything right now -- i just would like a hole for it left in the API.

ideally in the future it would cause a lookup of the named terminfo database, and translate according to that dictionary.

@tomek-szczesny

**Reply by dankamongmen** [reply]
but it's not incumbent on you to write that by any means. just leave the space in the API for it. if you want, check that it's always `NULL` for now even.

**Reply by tomek-szczesny** [reply]
Got it - I'll just ignore the input for now.

Mind you, I'm chilling in the mountains far from home, near Slovakian border, so I'll get back to all that next week. :)

**Reply by dankamongmen** [reply]
have an awesome time, mang! if you want to put up what you have before you go, i can review what's there. if you'd rather not, no worries!

**Reply by tomek-szczesny** [reply]
Thanks, I am visiting an old friend of mine. No mobile coverage, we steal some wifi from the neighbors, we carry water from a well, chop wood with axes, that kind of living. I promised help fixing some minor issues around the house - finally a bit of manual labor for my geeky ass.

The state of work is always available at [the proto repo](https://github.com/tomek-szczesny/notcurses-vt-proto). You may want to share your thoughts in the issues, or bash my code. ;) Keep in mind this code is evolving and probably it will get replaced in its entirety at some point, and that I am aware of many terrible practices present in there. If you still feel like reviewing it, be my guest!
